src/pydvl/influence/frameworks/torch_differentiable.py

```diff
@@ -2,36 +2,231 @@
+
+    :param hvp: a Callable Hvp, operating with tensors of size N
+    :param b: a vector or matrix, the right hand side of the equation $Hx = b$.
+    :param x0: initial guess for hvp
+    :param rtol: maximum relative tolerance of result
+    :param atol: absolute tolerance of result
+    :param maxiter: maximum number of iterations. If None, defaults to 10*len(b)
+    """
+    if x0 is None:
+        x0 = torch.clone(b)
+    if maxiter is None:
+        maxiter = len(b) * 10
+
+    y_norm = torch.sum(torch.matmul(b, b)).item()
+    stopping_val = max([rtol**2 * y_norm, atol**2])
+
+    x = x0
+    p = r = (b - hvp(x)).squeeze().type(torch.float64)
+    gamma = torch.sum(torch.matmul(r, r)).item()
+    optimal = False
+
+    for k in range(maxiter):
+        Ap = hvp(p).squeeze()
+        alpha = gamma / torch.sum(torch.matmul(p, Ap)).item()
+        x += alpha * p
+        r -= alpha * Ap
+        gamma_ = torch.sum(torch.matmul(r, r)).item()
+        beta = gamma_ / gamma
+        gamma = gamma_
+        p = r + beta * p
+
+        if gamma < stopping_val:
+            optimal = True
+            break
+
+    info = {"niter": k, "optimal": optimal}
+    return x, info
+
+
+def solve_lissa(
+    model: TwiceDifferentiable,
+    training_data: DataLoader,
+    b: torch.Tensor,
+    *,
+    lam: float = 0.0,
+    maxiter: int = 1000,
+    damp: float = 0.0,
+    scale: float = 10.0,
+    h0: Optional[torch.Tensor] = None,
+    rtol: float = 1e-4,
+    progress: bool = False,
+) -> torch.Tensor:
+    """
+    It uses LISSA, Linear time Stochastic Second-Order Algorithm, to calculate the
+    inverse of the HVP. More precisely, it finds x s.t. $Hx = b$, with $H$ being
+    the model hessian.
+    This is done by iteratively approximating H through
+    $$
+    H^{-1}_{j+1} b = b + (I - H) \ H^{-1}_j b
+    $$
+    where I is the identity matrix. Additional damping and scaling factors are
+    applied to help convergence. More info can be found in
+    :footcite:t:`koh_understanding_2017`
+
+    :param model: A model wrapped in the TwiceDifferentiable interface.
+    :param training_data: A DataLoader containing the training data.
+    :param b: a vector or matrix, the right hand side of the equation $Hx = b$.
+    :param lam: regularization of the hessian
+    :param progress: If True, display progress bars.
+    :param maxiter: maximum number of iterations,
+    :param damp: damping factor, defaults to 0 for no damping
+    :param scale: scaling factor, defaults to 10
+    :param h0: initial guess for hvp
+
+    :return: A matrix of shape [NxP] with each line being a solution of $Ax=b$.
+    """
+    if h0 is None:
+        h_estimate = torch.clone(b)
+    else:
+        h_estimate = h0
+    shuffled_training_data = DataLoader(
+        training_data.dataset, training_data.batch_size, shuffle=True
+    )
+    lissa_step = lambda h: b + (1 - damp) * h - reg_hvp(h) / scale  # type: ignore
```

Comment:
Could you please make this a proper function with a docstring? It would make the code more readable

Reply:
Should be done. Let me know if you feel it needs more docs